README.md
# tmle4rcts: TMLE for randomized trials

Code by Laura Balzer (<laura.balzer@berkeley.edu>) 

With edits and packaging by Josh Nugent (<joshuanugent@gmail.com>)
 
To install and run:

```
library(devtools)
devtools::install_github(repo = "https://github.com/LauraBalzer/tmle4rcts")
library(tmle4rcts)
```


changelog

Changes from "Meta" code shared on July 12. Hopefully I don't miss anything. -Josh

 - Arm-specific CV variance estimate changes:
  - Uses outputs DY0 and DY1 from get.IC.variance() in do.TMLE.validset()
  - passes those up through get.IC.CV(), where it collects them (think LOOCV) and calculates variance
  - then to CV.selector(), then to do.adaptive.prespec(), then Stage2(), which includes them as appropriate

 - made remove.pscore = NULL the default, can be overridden in Stage2() call. Sets TRUE if # ind units <=40.

 - Implemented sample.effect at Stage2() input and have ensured it cascades down to get.IC.variance()

 - Moved get.IC.variance() to utility_functions.R

 - Changed CV.selector() to output CV.risk = CV.risk[adj.index], not just CV.risk = CV.risk

 - Implemented various small input checks at start of Stage2()
  - Moved the code that checks that the alternative hypothesis direction is specified if a one-sided p-value is desired into the start of Stage2() rather than in get.inference to speed up the production of errors.
 
 - Fixed the get.cand.adj() to output ALL subsets and algorithms (not just all/one variable)
   - allows for low/medium/high complexity based on # of independent units per covariate, or you can user-specify
   - Made the function generic, not Q/g
   - Moved the call to the do.adaptive.prespec() function; calls it twice, removing selected Q variables for the second run for gAdj - changed how it removes variables for selecting the final gAdj
   - Moved get.cand.adj() to utility_functions.R for easier side-by-side editing
   - Commented out if(length(CAND.FORM) == 1) check from start of CV.selector(), should not be needed anymore

 - created pre-processing/data checking function in utility_functions.R
   - Gives warning message if data does not have 'id' column, and creates unique id for each row
   - Checks for "U" column, adds it if not there
   - Checks that data has a "Y" column
   - Adds alpha = 1 column if nothing there (and a warning)
   - If there is a "U" column in the data already but it's not all 1s, replaces it and generates a warning

 - Changed Stage2() default to be two-sided test
 - Added a lot of explicit return() statements for clarity
 - Updated est.df code in Stage2() to return the candidate adjustment variables used instead of their index numbers

 - Put %nin%, mean.harmonic, aggregage_IC, get.inference into utility_functions.R
 - Copied ltmle::HouseholdIC function, renamed it aggregate_IC
 - Created utility_functions.R for shared helper functions
 - Changed the R script names to be simpler, more general


R/test_functions.R
#' Test Me Out
#' @description A test function made while working out package and github setup.
#'
#' @param x The input object
#'
#' @return A character string.
#' @export
#'
#' @examples
#' test_me_out(1)
#' test_me_out("Yes")
test_me_out <- function(x){
  if(x == "Yes"){
    return("Yayyyyyyy")
  } else {
    return("Nooooooooooo")
  }
}

0_workflow.R
library(devtools)
library(tidyverse)
library(earth)
`%nin%` <- purrr::negate(`%in%`)

check() # then...
# commit and push to github



# write a function, use roxygen skeleton to document it
# COMMON ROXYGEN TAGS
# @description @examples @examplesIf
# @export @family @inheritParams @param
# @returns @seealso @rdname
# To wrap blocks to be less then 80 characters wide: 
#  use Ctrl/Cmd + Shift + /
#  Or menu: code -> re-flow comment

# To other documentation:
#   \code{\link{function}}: function in this package.
# \code{\link[MASS]{abbey}}: function in another package.
# \link[=dest]{name}: link to dest, but show name.
# \code{\link[MASS:abbey]{name}}: link to function in another package, but show name.
# \linkS4class{abc}: link to an S4 class.
# To the web:
#   \url{http://rstudio.com}: a url.
# \href{http://rstudio.com}{Rstudio}:, a url with custom link text.
# \email{hadley@@rstudio.com} (note the doubled @): an email address.

# formatting with LATEX-style...
# \emph{will give italics}
# \strong{will give bold}
# \code{inline code format}
# \preformatted{for multiline code}
# \itemize{
    # \item First thing
    # \item Second thing
# }
# \tabular, etc.

# Special characters:
#   Use @@ to get @
#   Use \% to get %
#   Use \\ to get \

# use @inheritParams AnotherFunctionWithASharedArgument
# to copy parameters from that other function (not recursively tho)

# ... then...
devtools::document()
check()

################################################################## commit and push to GH again


##### TRY INSTALLING FROM GITHUB
library(devtools)
devtools::install_github(repo = "https://github.com/joshua-nugent/tmle4rcts")
library(tmle4rcts)
#library(tidyverse)
?test_me_out
?get.inference
?get.IC.variance
?Stage2
?do.TMLE
?do.data.adapt



################ OR...
# skip it and just install freshest version
install()
library(tmle4rcts)
dat <- simulate_clustered_data()
Stage2(data.input = dat, goal = "RD",
       psi = dat$unit_mean_diff[1], remove.pscore = F,
       do.data.adapt = T, verbose = F, sample.effect = F,
       cand.Qform = c("glm"),
       cand.gform = c("glm"),
       cand.QAdj = c("covar0", "covar1", "covar2"),
       cand.gAdj = c("covar0", "covar1", "covar2"))
Stage2(data.input = dat, goal = "aRR",
       psi = dat$unit_mean_diff[1], remove.pscore = T,
       do.data.adapt = T, verbose = F, sample.effect = F,
       cand.Qform = c("glm", "lasso"),
       cand.gform = c("glm","step","mars"),
       cand.QAdj = c("covar0", "covar1", "covar2"),
       cand.gAdj = c("covar0", "covar1", "covar2"))
head(dat)




#combs <- do.call(c, lapply(seq_along(ccs), combn, x = ccs, simplify = FALSE))
#combs
(combs <- do.call(c, lapply(seq_along(1:length(ccs)), combn, x = 1:length(ccs), simplify = FALSE)))
length(combs)
dim(expand.grid(forms = forms, combs = combs))
expand.grid(forms = forms, combs = combs)
forms[expand.grid(forms = forms, combs = combs)[[14,"forms"]]]
ccs[expand.grid(forms = forms, combs = combs)[[14,"combs"]]]


devtools::install_github(repo = "https://github.com/joshua-nugent/tmle4rcts")
library(tmle4rcts)
forms <- c("glm","step", "lasso")
ccs <- c("a", "b", "c")
get.cand.adj(cand.vars = ccs, cand.algos = forms, data.adapt.complexity = "high")
get.cand.adj(cand.vars = ccs, cand.algos = forms, data.adapt.complexity = "med")
get.cand.adj(cand.vars = ccs, cand.algos = forms, data.adapt.complexity = "low")




install()
dat <- simulate_clustered_data(treatment_arm_clusters = 10,
                               control_arm_clusters = 10)
library(tmle4rcts)
Stage2(data.input = dat, goal = "RD",
       psi = dat$unit_mean_diff[1], remove.pscore = F,
       do.data.adapt = T, verbose = F, sample.effect = F, do.cv.variance = T,
       cand.Qform = c("glm"), cand.gform = c("glm"),
       cand.QAdj = c("covar0", "covar2"),
       cand.gAdj = c("covar0", "covar2"))
Stage2(data.input = dat, goal = "RD",
       psi = dat$unit_mean_diff[1], remove.pscore = F,
       do.data.adapt = T, verbose = F, sample.effect = F, do.cv.variance = F,
       cand.Qform = c("glm"), cand.gform = c("glm"),
       cand.QAdj = c("covar0", "covar2"),
       cand.gAdj = c("covar0", "covar2"))

dat <- simulate_clustered_data(treatment_arm_clusters = 30,
                               control_arm_clusters = 30)
library(tmle4rcts)
Stage2(data.input = dat, goal = "RD",
       psi = dat$unit_mean_diff[1], remove.pscore = F,
       do.data.adapt = T, verbose = F, sample.effect = F, do.cv.variance = T,
       cand.Qform = c("glm"), cand.gform = c("glm"),
       cand.QAdj = c("covar0", "covar2"),
       cand.gAdj = c("covar0", "covar2"))



















R/data_generators.R




simulate_clustered_data <- function(treatment_arm_cluster_size = 15, treatment_arm_clusters = 50,
                                    control_arm_cluster_size = 5, control_arm_clusters = 50,
                                    txt_eff = .5, ranef_sd = 1, covar1_coef = 1, seed = NA,
                                    informative_cluster_size = F){
  if(!is.na(seed)){ set.seed(seed) }
  total_units_txt <- treatment_arm_cluster_size*treatment_arm_clusters
  total_units_ctr <- control_arm_cluster_size*control_arm_clusters
  total_units <- total_units_txt + total_units_ctr
  unit_id <- 1:(total_units)
  cluster_id_txt <- rep(1:treatment_arm_clusters, each = treatment_arm_cluster_size)
  cluster_id_ctr <- rep((max(cluster_id_txt)+1):(max(cluster_id_txt) + control_arm_clusters), each = control_arm_cluster_size)
  cluster_id <- c(cluster_id_txt, cluster_id_ctr)
  total_clusters <- max(cluster_id)
  cluster_lengths <- c(rep(treatment_arm_cluster_size, times = treatment_arm_clusters),
                       rep(control_arm_cluster_size, times = control_arm_clusters))
  cluster_cov <- sd(cluster_lengths) / mean(cluster_lengths)
  N_j <- rep(cluster_lengths, times = cluster_lengths)
  
  A <- c(rep(1, times = total_units_txt), rep(0, times = total_units_ctr))
  
  if(ranef_sd == 0){
    UE <- 0
  } else {
    UEs <- exp(rnorm(n = total_clusters, mean = 0, sd = ranef_sd))
    UE <- rep(UEs, times = cluster_lengths)
  }
  UY <- runif(n = total_units, min = 0, max = 2)
  covar0 <- rnorm(n = total_units, mean = UE, sd = .5)
  covar1 <- rbinom(n = total_units, size = 1, prob = .2)
  covar2 <- rnorm(n = total_units, mean = 0, sd = .5)
  covar3 <- rnorm(n = total_units, mean = 0, sd = .5)
  
  Y0 <- covar1_coef*covar1 + covar0^2 + covar0*covar1 + UY + UE
  Y1 <- covar1_coef*covar1 + covar0^2 + covar0*covar1 + UY + UE + txt_eff
  if(informative_cluster_size){
    Y1 <- Y1 + txt_eff*N_j/total_units
  }
  Y <- ifelse(A == 1, Y1, Y0)
  dat <- cbind.data.frame(unit_id, cluster_id, A,
                          UY, covar0, covar1, covar2, covar3, UE,
                          txt_eff, Y0, Y1, Y,
                          N_j = N_j) %>% dplyr::mutate(id = cluster_id, alpha = 1, U = 1)
  cluster_means <- dat %>% dplyr::group_by(cluster_id, A) %>% dplyr::summarise(cluster_mean0 = mean(Y0),
                                                                               cluster_mean1 = mean(Y1),
                                                                               cluster_meanY = mean(Y))
  ind_weight <- total_clusters / total_units
  ind_weighted_cluster_means <- dat %>% dplyr::group_by(cluster_id, A) %>%
    dplyr::summarise(ind_weighted_cluster_mean0 = sum(Y0)*ind_weight,
                     ind_weighted_cluster_mean1 = sum(Y1)*ind_weight,
                     ind_weighted_cluster_meanY = sum(Y)*ind_weight)
  cluster_mean_diff <- mean(cluster_means$cluster_mean1) - mean(cluster_means$cluster_mean0)
  cluster_mean_irr <- mean(cluster_means$cluster_mean1) / mean(cluster_means$cluster_mean0)
  ind_weighted_cluster_mean_diff <- mean(ind_weighted_cluster_means$ind_weighted_cluster_mean1) - mean(ind_weighted_cluster_means$ind_weighted_cluster_mean0)
  ind_weighted_cluster_mean_irr <- mean(ind_weighted_cluster_means$ind_weighted_cluster_mean1) / mean(ind_weighted_cluster_means$ind_weighted_cluster_mean0)
  
  return(cbind.data.frame(dat,
                          covar1_coef = covar1_coef,
                          true_rd_input = txt_eff,
                          true_covar_coef_input = covar1_coef,
                          ranef_sd_input = ranef_sd,
                          cluster_cov = cluster_cov,
                          sample_mean0 = mean(Y0),
                          sample_mean1 = mean(Y1),
                          unit_mean_diff = mean(Y1) - mean(Y0),
                          unit_irr = mean(Y1) / mean(Y0),
                          cluster_mean_diff = cluster_mean_diff,
                          cluster_mean_irr = cluster_mean_irr,
                          ind_weighted_cluster_mean_diff = ind_weighted_cluster_mean_diff,
                          ind_weighted_cluster_mean_irr = ind_weighted_cluster_mean_irr
  ))
}


R/aps_functions.R




#' Implement adaptive pre-specification
#'
#' @description function to implement adaptive prespecification as described in
#'   Balzer et al. "Adaptive pre-specification in randomized trials with and
#'   without pair-matching"
#'
#' @param goal goal of analysis ('aRR', 'OR', 'RD')
#' @param target Target of estimation and inference: cluster-level ("clust") or
#'   pooled-indv effect ("indv")
#' @param break.match Logical indicating whether to break matched pairs (if they
#'   exist)
#' @param Ldata Input data set
#' @param V
#' @param cand.QAdj Character vector of names of candidate adjustment variables
#'   for the outcome regression
#' @param cand.Qform Character vector of names of candidate adjustment
#'   algorithms for the outcome regression
#' @param cand.gAdj Character vector of names of candidate adjustment variables
#'   for the propensity score
#' @param cand.gform Character vector of names of candidate adjustment
#'   algorithms for the propensity score
#' @param remove.pscore if T, remove the variable(s) selected for adjustment in
#'   the outcome regression from candidates for the pscore... should only be
#'   used if doing adaptive prespec in RCT with few indpt units
#' @param QAdj selected adjustment variable for the outcome regression, if
#'   specifying in advannce
#' @param gAdj selected adjustment variable for the pscore regression, if
#'   specifying in advance
#' @param scale_value maximum value for outcome scaling
#' @param scale_value_min minimum value for outcome scaling
#' @param verbose
#' @param sample.effect If the target of inference the sample effect or the
#'   population effect?
#' @param data.adapt.complexity Set to 'low', 'med', or 'high'. See
#'   ?get.cand.adj for details.
#'
#' @return Working model selection for candidate TMLE
#' @export
#' 
#' @examples
do.adaptive.prespec <- function(goal, target='indv', break.match=T, Ldata, V=5,
                                cand.QAdj, cand.Qform,
                                cand.gAdj, cand.gform,
                                remove.pscore, QAdj = NULL, gAdj = NULL,
                                scale_value, scale_value_min, verbose = F,
                                sample.effect, data.adapt.complexity){
  
  # get the indpt units (will be each observation in and individual-level RCT)
  if( !break.match ){
    Ldata$indpt.unit <- Ldata$pair
  } else {
    Ldata$indpt.unit <- Ldata$id
  }
  unique.ids <- unique(Ldata$indpt.unit)
  
  # get folds
  if(length(unique.ids) > 40){
    # V-fold CV
    folds <- get.folds(V=V, Y=Ldata$Y, ids=unique.ids)
  } else {
    # leave-one-out CV
    folds <- vector("list", length(unique.ids))
    for(v in seq(length(unique.ids))){
      folds[[v]] <- unique.ids[v]
    }
  }
  
  if(is.null(remove.pscore)){
    if(length(unique.ids) > 40){
      remove.pscore <- F
    } else {
      remove.pscore <- T
    }
    message(paste0("remove.pscore not speficied; setting to ", remove.pscore, " based on # of independent units"))
  }

  
  #=====================================================
  # GENERATE COVARIATE SET / WORKING MODEL COMBINATIONS FOR OUTCOME REGRESSION
  # Figure out complexity level first
  if(is.null(data.adapt.complexity)){
    nvars <- length(union(cand.gAdj, cand.QAdj))
    ratio <- length(unique.ids) / nvars
    low_cutoff <- 10
    med_cutoff <- 50
    if(ratio < low_cutoff){
      message(paste0("Fewer than ", low_cutoff, " independent units per covariate. Setting data.adapt.complexity = 'low'"))
      data.adapt.complexity <- "low"
    } else if (ratio < med_cutoff){
      message(paste0("Between ", low_cutoff, " and ", med_cutoff, " independent units per covariate. Setting data.adapt.complexity = 'med'"))
      data.adapt.complexity <- "med"
    } else {
      message(paste0("More than ", med_cutoff, " independent units per covariate. Setting data.adapt.complexity = 'high'"))
      data.adapt.complexity <- "high"
    }
  }
  cand.Q <- get.cand.adj(cand.vars = cand.QAdj, cand.algos = cand.Qform,
                         data.adapt.complexity = data.adapt.complexity)
  cand.QAdj <- cand.Q$cand.varset
  cand.Qform <- cand.Q$cand.algos


  
  
  #=====================================================
  # FIGURE OUT WHICH COMBINATION FOR OUTCOME REGRESSION IS BEST
  if( is.null(QAdj) ){
    
    if(verbose) print("Examining sets for outcome regression")
    
    # do adaptive pre-specification to select from candidate approaches for Qbar
    select.Q <- suppressWarnings(CV.selector(goal=goal, target=target, break.match=break.match, Ldata=Ldata,
                                              CAND.ADJ = cand.QAdj, CAND.FORM=cand.Qform, forQ=T, verbose = verbose,
                                              scale_value=scale_value, scale_value_min=scale_value_min,
                                              folds=folds, sample.effect = sample.effect
                                              ))
    #if(verbose) print(select.Q)
    Q.index <- select.Q$adj.index
    QAdj <- select.Q$Adj
    Qform <- select.Q$form
    #if(verbose){print("Selected for Q:") print(QAdj); print(Qform)}
    
    # if select unadjusted estimator for QbarAW=E(Y|A,W), then stop
    if(sum('U' %in% QAdj) == 1){ 
      g.index <- -99; gAdj <- 'U'; gform <- 'glm'
      var.CV <- select.Q$var.CV
    }
    
    if((sum(QAdj %in% 'U')) == 0 & remove.pscore){
        if(verbose) print('remove.pscore = T; removing selected QAdj variable(s) from candidates for gAdj')
        cand.gAdj <- cand.gAdj[cand.gAdj %nin% QAdj]
        if(length(cand.gAdj) == 0){
          if(verbose) print("all covars used for outcome regression. setting cand.gAdj = U")
          cand.gAdj <- "U"
        }
    }
  }

  
  #=====================================================
  # GENERATE COVARIATE SET / WORKING MODEL COMBINATIONS FOR PROPENSITY SCORE
  cand.g <- get.cand.adj(cand.vars = cand.gAdj, cand.algos = cand.gform,
                         data.adapt.complexity = data.adapt.complexity)
  cand.gAdj <- cand.g$cand.varset
  cand.gform <- cand.g$cand.algos


  
  #=====================================================
  # FIGURE OUT WHICH COMBINATION FOR PROPENSITY SCORE IS BEST
  if( is.null(gAdj) ){ 		
    if(verbose) print("Examining sets for propensity score")
    
    select.G <- suppressWarnings(CV.selector(goal = goal, target = target,
                                              break.match = break.match, Ldata = Ldata,
                                              CAND.ADJ = cand.gAdj, CAND.FORM = cand.gform,
                                              forQ = F,  verbose = verbose,
                                              # input selected variables/form of the outcome regression
                                              QAdj = QAdj, Qform = Qform,
                                              scale_value = scale_value, scale_value_min = scale_value_min,
                                              folds = folds,
                                              sample.effect = sample.effect))
    
    g.index <- select.G$adj.index
    gAdj <- select.G$Adj
    gform <- select.G$form
    var.CV <- select.G$var.CV		
    var.CV.1 <- select.G$var.CV.0		
    var.CV.0 <- select.G$var.CV.0		
    #if(verbose){print(select.G);    print("Selected for g:");  print(gAdj);  print(gform)}
  }		
  output_aps <- list(Q.index = Q.index, QAdj = QAdj, Qform = Qform, 
                     g.index = g.index, gAdj = gAdj, gform = gform,
                     var.CV = var.CV, var.CV.1 = var.CV.1, var.CV.0 = var.CV.0)
  #if(verbose) print(output_aps)
  return(output_aps)
}



#-----------------------------------------------------#-----------------------------------------------------
# CV.selector: function to estimate the cross-validated risk
#		Loss function is the squared-IC; Risk is then the variance of the TMLE
#		See Balzer et al. "Adaptive pre-specification in randomized trials with and without pair-matching"
#
#	input: goal of analysis ('aRR' or 'RD), 
#   target of inference 
#		indicator to break the match (break.match)
#		dataset (Ldata)
#		candidate adjustment variables; they do not have to be at the cluster-level
#		indicator if for the conditional mean outcome (forQ)
#		selected adjustment variable for the outcome regression (QAdj)
#	output: selection for adjustment variable (corresponding to a TMLE)
#-----------------------------------------------------#-----------------------------------------------------


CV.selector <- function(goal, target, break.match, Ldata, CAND.ADJ, CAND.FORM, 
                        forQ, QAdj=NULL, Qform=NULL, verbose,
                        scale_value, scale_value_min, folds, sample.effect){
  # if(length(CAND.FORM) == 1){
  #   # if exploring only one estimation algorithm (usually GLM) then need to replicate the number forms
  #   CAND.FORM <- rep(CAND.FORM, length(CAND.ADJ))
  # }
  
  if(length(CAND.FORM) != length(CAND.ADJ)){ # After changes to get.cand.adj, this should not happen.
    stop('PROBLEM: MISMATCH LENGTHS OF ADJ VAR SETS AND MODELS')
  }

  # Number of candidate estimators is given by length Qform//gform
  num.tmles <- length(CAND.FORM)
  CV.risk <- var.CV <- var.CV.1 <- var.CV.0 <- rep(NA, num.tmles)
  
  for(k in 1: num.tmles){	
    
    if(forQ){
      # print("evaluating a Q option:")
      # print(CAND.ADJ[[k]])
      # print(CAND.FORM[k])
      # if selecting the adjustment approach for the outcome regression
      IC.temp <- get.IC.CV(goal = goal, target = target, break.match = break.match, Ldata = Ldata,
                           QAdj = CAND.ADJ[[k]], Qform = CAND.FORM[k],
                           gAdj = NULL, gform = 'glm', verbose = verbose,
                           scale_value = scale_value, scale_value_min = scale_value_min, 
                           folds = folds, sample.effect = sample.effect)
    } else {
      # print("evaluating g option:")
      # print(CAND.ADJ[[k]])
      # print(CAND.FORM[k])
      # if collaboratively selecting the adjustment approach for the pscore
      IC.temp <- get.IC.CV(goal = goal, target = target, break.match = break.match, Ldata = Ldata, 
                           QAdj = QAdj, Qform = Qform, verbose = verbose,
                           gAdj = CAND.ADJ[[k]], gform = CAND.FORM[k],
                           scale_value = scale_value, scale_value_min = scale_value_min, 
                           folds = folds, sample.effect = sample.effect)
    }
    # if(verbose) print(IC.temp)
    # estimating the CV risk for each candidate
    CV.risk[k]<- IC.temp$CV.risk
    # estimating the CV variance for that TMLE
    var.CV[k] <- IC.temp$var.CV
    var.CV.1[k] <- IC.temp$var.CV.1
    var.CV.0[k] <- IC.temp$var.CV.0
  }
  #print(CV.risk)

  # select the candidate estimator resulting in the smallest CV-risk
  adj.index <- which.min(CV.risk)
  output_best <- list(CV.risk = CV.risk[adj.index],
                      adj.index = adj.index, Adj = CAND.ADJ[[adj.index]], form = CAND.FORM[adj.index],
                      var.CV = var.CV[adj.index],
                      var.CV.1 = var.CV.1[adj.index],
                      var.CV.0 = var.CV.0[adj.index])
  return(output_best)
}




#' Calculate a cross-validated estimate of the influence curve 
#'
#' UPDATES
#' previous version only did leave-one-out (unit or pair)
#' this version generalizes to V-fold CV if V>=40
#' - can input the number of folds V (default=10)
#' - folds created stratified on binary outcomes (by default)
#' - if stratify=T and # observations in a given class is <V, 
#' then sets V=min observations in that fold
#'
#' See Balzer et al. "Adaptive pre-specification in randomized trials with and without pair-matching."
#'
#' @param goal goal of analysis ('aRR' or 'RD)
#' @param target cluster/individual effect (JN: Not implemented yet?)
#' @param break.match indicator to break the match
#' @param Ldata input dataset
#' @param QAdj adjustment variable for the outcome regression
#' @param Qform adjustment approach for outcome regression
#' @param gAdj adjustment variable for the pscore
#' @param gform adjustment approach for pscore regression
#' @param scale_value For outcomes not bounded within [0,1], the maximum value of the outcome.
#' @param scale_value_min For outcomes not bounded within [0,1], the minimum value of the outcome.
#' @param folds Number of folds in cross-validation.
#'
#' @return cross-validated estimate of the IC for pair
#' @export
#'
#' @examples
get.IC.CV <- function(goal, target, break.match, verbose,
                      Ldata, QAdj, Qform, gAdj=NULL, gform='glm', 
                      scale_value, scale_value_min, folds, sample.effect){
  
  nFolds <- length(folds)
  DY.CV <- CV.risk <- DY1.CV <- DY0.CV <- NULL
  
  # doing a cross-validated estimate
  for(i in 1:nFolds) {
    
    these <- Ldata$indpt.unit %in% folds[[i]]  ########  IMPORTANT!!!!
    valid <- Ldata[these, ]
    train <- Ldata[!these,]
    
    # run full TMLE algorithm on the training set
    train.out <- do.TMLE(goal=goal, target=target, train=train, QAdj=QAdj, Qform=Qform, 
                         gAdj=gAdj, gform=gform,
                         scale_value=scale_value, scale_value_min=scale_value_min,
                         doing.CV=T, verbose=F, sample.effect = sample.effect)	
    
    # get the relevant components of the IC for the validation set, 
    # using fits based on the training set
    valid.out <- do.TMLE.validset(goal=goal, target=target, valid=valid, train.out=train.out,
                                  scale_value=scale_value, scale_value_min=scale_value_min,
                                  sample.effect = sample.effect)	

    # estimating the CV risk for each candidate
    # risk = Expectation of loss with loss as IC-sq
    # risk = variance of TMLE
    if(break.match){
      DY.CV <- c(DY.CV, valid.out$DY)
      CV.risk <- c(CV.risk, mean(valid.out$DY^2))
    } else {
      DY.CV <- c(DY.CV, valid.out$DY.paired)
      CV.risk <- c(CV.risk, mean(valid.out$DY.paired^2))
    }
    DY1.CV <- c(DY1.CV, valid.out$DY1)
    DY0.CV <- c(DY0.CV, valid.out$DY0)
  }
  #if(verbose){print(valid.out)}
  
  # average across folds
  CV.risk <- mean(CV.risk)
  
  # estimating the CV variance for that TMLE
  var.CV <- stats::var(DY.CV) / length(DY.CV)
  var.CV.1 <- stats::var(DY1.CV) / length(DY1.CV)
  var.CV.0 <- stats::var(DY0.CV) / length(DY0.CV)
  
  return(list(CV.risk = CV.risk, var.CV = var.CV,
              var.CV.0 = var.CV.0, var.CV.1 = var.CV.1))
}



#-----------------------------------------------------#-----------------------------------------------------
# do.TMLE.for.valid: function to obtain a cross-validated estimate of the influence curve
#	for observations in the validation set
#		See Balzer et al. "Adaptive pre-specification in randomized trials with and without pair-matching"
#
#	input: goal of analysis ('aRR' or 'RD'),
#		validation dataset ('valid') 
#		TMLE-fits from training set (train.out)
#	output: cross-validated estimate of the IC,
#		cross-validated risk estimate (loss=IC^2)
#-----------------------------------------------------#-----------------------------------------------------

do.TMLE.validset <- function(goal, target, valid,
                             train.out, scale_value, scale_value_min,
                             sample.effect){
	
	# J <- length(unique(valid$id) )

	#=============================================
	# Step1 - initial estimation of E(Y|A,W)= Qbar(A,W)
	#=============================================
	valid <- do.Init.Qbar(train=valid, QAdj=train.out$QAdj, Qform=train.out$Qform, 
	                     glm.out=train.out$Q.out)$train
	
	#=============================================
	# Step2: Calculate the clever covariate
	#=============================================
	valid <- get.clever.cov(train=valid, gAdj=train.out$gAdj, gform=train.out$gform, 
	                        p.out=train.out$p.out)$train

	#=============================================
	# Step3: Targeting - 			
	#=============================================
	valid <- do.targeting(train=valid, eps=train.out$eps, goal=goal)
	
	#=============================================
	# Step5: Variance estimation using treatment-specific means from training set
	#=============================================
	 get.IC.variance(goal=goal, target=target, Vdata=valid, R1=train.out$R1, R0=train.out$R0, 
	                 scale_value=scale_value, scale_value_min=scale_value_min,
	                 doing.CV=T, sample.effect = sample.effect)
}








#-----------------------------------------------------#-----------------------------------------------------

# adapted from .cvFolds from cvAUC package: https://CRAN.R-project.org/package=cvAUC
# by Erin LeDell 
get.folds <- function(V, Y, ids, stratify=T){
  
  if(stratify & length(unique(Y))==2){
    # stratify on the outcome
    classes <- tapply(1:length(Y), INDEX=Y, FUN=split, 1)
    ids.Y1 <- ids[classes$`1`]
    ids.noY1 <- ids[classes$`0`]
    if(length(ids.Y1) < V | length(ids.noY1) < V) {
      V <- min( length(ids.Y1), length(ids.noY1))
    }
    ids.Y1.split <- split(sample(length(ids.Y1)), rep(1:V, length=length(ids.Y1)))
    ids.noY1.split <- split(sample(length(ids.noY1)), rep(1:V, length=length(ids.noY1)))
    folds <- vector("list", V)
    for (v in seq(V)){
      folds[[v]] <- c(ids.Y1[ids.Y1.split[[v]]], ids.noY1[ids.noY1.split[[v]]])
    }
    
  } else {
    # dont stratify on the outcome
    ids.split <- split(sample(length(ids)), rep(1:V, length=length(ids)))
    folds <- vector("list", V)
    for (v in seq(V)){
      folds[[v]] <- ids[ids.split[[v]]]
    }
    
  }
  folds
}
